src/Data/Generic/Rep/Enum.purs

```diff
@@ -0,0 +1,145 @@
+module Data.Generic.Rep.Enum
+  ( genericSucc
+  , class GenericSucc
+  , succ'
+
+  , genericPred
+  , class GenericPred
+  , pred'
+
+  , genericFromEnum
+  , class GenericFromEnum
+  , fromEnum'
+
+  , genericToEnum
+  , class GenericToEnum
+  , toEnum'
+
+  , genericCardinality
+  , class GenericCardinality
+  , cardinality'
+  ) where
+
+import Data.Generic.Rep
+import Data.Maybe (Maybe(..))
+import Data.Enum (Cardinality(Cardinality))
+import Data.Newtype (unwrap)
+import Prelude ((<$>), (+), (-))
+
+class GenericSucc a where
+  succ' :: a -> Maybe a
+
+instance genericSuccSumConstructor ::
```

Comment:
Definitely agree that the types involved here are somewhat ugly. I will give this another, more thorough look tomorrow and try to come up with something better. In the meantime, though, some questions:

Why no instance for `NoConstructors`: I couldn't see how such a thing would make sense. If we wrote one, would it simply always return `Nothing`? Since there are no constructors for the data type, how would `succ` get called in the first place? I suspect there might be a gap in my understanding of constructor-less data types here. Can you show me an example of one that makes sense as an instance of `Enum`?

Likewise, I couldn't see how an instance for `Constructor` would make sense, either, as a data type with `Cardinality 1` would have `succ a == pred a` (both `Nothing`), in violation of the typeclass laws stated in the documentation.